File: src/sp/it/pl/util/file/FileMonitor.java
package sp.it.pl.util.file;

import java.io.File;
import java.io.IOException;
import java.nio.file.ClosedWatchServiceException;
import java.nio.file.FileSystems;
import java.nio.file.Path;
import java.nio.file.WatchEvent;
import java.nio.file.WatchEvent.Kind;
import java.nio.file.WatchKey;
import java.nio.file.WatchService;
import java.util.concurrent.ThreadFactory;
import java.util.function.BiConsumer;
import java.util.function.Consumer;
import java.util.function.Predicate;
import sp.it.pl.util.async.executor.EventReducer;
import sp.it.pl.util.system.Os;
import static com.sun.nio.file.ExtendedWatchEventModifier.FILE_TREE;
import static java.nio.file.StandardWatchEventKinds.ENTRY_CREATE;
import static java.nio.file.StandardWatchEventKinds.ENTRY_DELETE;
import static java.nio.file.StandardWatchEventKinds.ENTRY_MODIFY;
import static java.nio.file.StandardWatchEventKinds.OVERFLOW;
import static sp.it.pl.util.async.AsyncKt.runFX;
import static sp.it.pl.util.async.AsyncKt.threadFactory;
import static sp.it.pl.util.dev.Util.logger;

// TODO: use interface, leverage recursive directory watch on Windows
// TODO: provide documentation
public class FileMonitor {
	/**
	 * Relatively simple to monitor a file? Think again.
	 * 1) Dir only!
	 * WatchService allows us to only monitor a directory. We then must simply ignore other
	 * events of files other than the one we monitor. This is really bad if we want to monitor
	 * multiple files in a single directory independently.
	 * Solvable using a predicate parameter to filter out unwanted events.
	 * 2) Modification events.
	 * Events can fire multiple times when application use safe-rewrite saving or in various other scenarios! E.g.
	 * editing and saving .java file in Netbeans fires 3 events at about 8-13 ms time gap (tested on SSD).
	 * Solvable using event reducer & checking modification times
	 * 3) Nested events
	 * CREATE and DELETE events are fired for files and directories up to level 2 (children of
	 * children of the monitored directory, e.g.,  mon_dir/lvl1/file.txt). MODIFIED events will
	 * be thrown for any direct child.
	 * Furthermore, recursive monitoring is only supported on Windows, this is a platform limitation.
	 */

	private static final ThreadFactory threadCreator = threadFactory("WatchServiceConsumer", true);
	private File monitoredFileDir;
	private WatchService watchService;
	private Predicate<File> filter;
	private BiConsumer<Kind<Path>,File> action;

	private final EventReducer<Event> modificationReducer = EventReducer.toLast(50, e -> emitEvent(e.kind, e.file));

	private void emitEvent(Kind<Path> type, File file) {
		// This works as it should.
		// If anyone needs logging, they are free to do so in the event handler.
		// log(FileMonitor.class).info("{} event {} on {}", name,type,file);

		// always run on fx thread
		runFX(() -> action.accept(type, file));
	}

	/**
	 * Creates and starts directory monitoring reporting events for single file contained within the directory.
	 * <p/>
	 * Shortcut for:
	 * <p/>
	 * {@code monitorDirsFiles(monitoredFile.getParentFile(), file -> file.equals(monitoredFile), (type,file) ->
	 * handler.accept(type)); }
	 * <p/>
	 *
	 * @param handler handles the event taking the event type as parameter
	 * @return directory monitor
	 */
	public static FileMonitor monitorFile(File monitoredFile, Consumer<Kind<Path>> handler) {
		return monitorDirsFiles(monitoredFile.getParentFile(), file -> file.equals(monitoredFile), (type, file) -> handler.accept(type));
	}

	/**
	 * Creates and starts directory monitoring for specified directory reporting events for any 1st
	 * level child file which passes the predicate filter.
	 *
	 * @param monitoredDir directory to be monitored
	 * @param filter filter narrowing down events, for example any text file.
	 * @param handler handles the event containing type and modified file parameters
	 * @return directory monitor
	 */
	@SuppressWarnings({"unchecked", "ConstantConditions"})
	public static FileMonitor monitorDirsFiles(File monitoredDir, Predicate<File> filter, BiConsumer<Kind<Path>,File> handler) {

		FileMonitor fm = new FileMonitor();
		fm.monitoredFileDir = monitoredDir;
		fm.filter = filter;
		fm.action = handler;

		Path dir = fm.monitoredFileDir.toPath();
		try {
			fm.watchService = FileSystems.getDefault().newWatchService();
			dir.register(fm.watchService, ENTRY_CREATE, ENTRY_DELETE, ENTRY_MODIFY, OVERFLOW);
			threadCreator.newThread(() -> {
				// The check requires I/O so lets do that on bgr thread as well
//                if (!monitoredDir.isDirectory()) throw new IllegalArgumentException("File not a directory or does not exist.");

				boolean valid;
				WatchKey watchKey;
				do {
					try {
						watchKey = fm.watchService.take();
					} catch (InterruptedException e) {
						logger(FileMonitor.class).error("Interrupted monitoring of directory {}", dir, e);
						return;
					} catch (ClosedWatchServiceException e) {
						// watching ended
						return;
					}

					for (WatchEvent<?> event : watchKey.pollEvents()) {
						Kind<?> type = event.kind();

						if (type==OVERFLOW) continue;

						WatchEvent<Path> ev = (WatchEvent<Path>) event;
						String modifiedFileName = ev.context().toString();
						File modifiedFile = new File(fm.monitoredFileDir, modifiedFileName);

//                        if (modifiedFile.getParentFile().equals(fm.monitoredFileDir)) // to do
						if (fm.filter.test(modifiedFile)) {
							if (type==ENTRY_MODIFY) {
								runFX(() ->
										fm.modificationReducer.push(new Event((Kind) type, modifiedFile))   // TODO: fix overwriting events
								);
							} else {
								fm.emitEvent((Kind) type, modifiedFile);
							}
						}
					}

					valid = watchKey.reset();
				} while (valid);
			}).start();
		} catch (IOException e) {
			logger(FileMonitor.class).error("Error when starting file monitoring {}", fm.monitoredFileDir, e);
		}

		return fm;
	}

	@SuppressWarnings({"unchecked", "ConstantConditions", "UnnecessaryLocalVariable"})
	public static FileMonitor monitorDirectory(File toMonitor, boolean recursive, BiConsumer<Kind<Path>,File> handler) {
		FileMonitor fm = new FileMonitor();
		fm.monitoredFileDir = toMonitor;
		fm.action = handler;

		Path dir = fm.monitoredFileDir.toPath();
		try {
			fm.watchService = FileSystems.getDefault().newWatchService();

			boolean recursiveRequested = recursive;
			boolean recursiveSupported = Os.WINDOWS.isCurrent();
			boolean recursiveWarn = recursiveRequested && !recursiveSupported;
			boolean recursiveUsed = !recursiveRequested || !recursiveSupported;

			if (recursiveWarn)
				logger(FileMonitor.class).warn("Recursive file watcher is not supported, using standard, file={}", fm.monitoredFileDir);

			if (recursiveUsed) dir.register(fm.watchService, ENTRY_CREATE, ENTRY_DELETE, ENTRY_MODIFY, OVERFLOW);
			else dir.register(fm.watchService, new Kind<?>[] {ENTRY_CREATE, ENTRY_DELETE, ENTRY_MODIFY, OVERFLOW}, FILE_TREE);

			threadCreator.newThread(() -> {
				// The check requires I/O so lets do that on bgr thread as well
//                if (!toMonitor.isDirectory()) throw new IllegalArgumentException("File not a directory or does not exist.");

				boolean valid;
				WatchKey watchKey;
				do {
					try {
						watchKey = fm.watchService.take();
					} catch (InterruptedException e) {
						logger(FileMonitor.class).error("Interrupted monitoring of directory {}", dir, e);
						return;
					} catch (ClosedWatchServiceException e) {
						// watching ended
						return;
					}

					for (WatchEvent<?> event : watchKey.pollEvents()) {
						Kind<?> type = event.kind();
						if (type!=OVERFLOW) {
							WatchEvent<Path> ev = (WatchEvent<Path>) event;
							String modifiedFileName = ev.context().toString();
							File modifiedFile = new File(fm.monitoredFileDir, modifiedFileName);

							fm.emitEvent((Kind) type, modifiedFile);
						}
					}

					valid = watchKey.reset();
				} while (valid);
			}).start();
		} catch (IOException e) {
			logger(FileMonitor.class).error("Error when starting directory monitoring {}", fm.monitoredFileDir, e);
		}

		return fm;
	}

	public void stop() {
		try {
			if (watchService!=null) watchService.close();
		} catch (IOException e) {
			logger(FileMonitor.class).error("Error when closing file monitoring {}", monitoredFileDir, e);
		}
	}

	private static class Event {
		public final Kind<Path> kind;
		public final File file;

		private Event(Kind<Path> kind, File file) {
			this.kind = kind;
			this.file = file;
		}
	}
}

File: src/util/main/sp/it/util/file/FileMonitor.kt
package sp.it.util.file

import com.sun.nio.file.ExtendedWatchEventModifier.FILE_TREE
import mu.KLogging
import sp.it.util.async.executor.EventReducer
import sp.it.util.async.executor.FxTimer.Companion.fxTimer
import sp.it.util.async.runFX
import sp.it.util.async.runNew
import sp.it.util.async.threadFactory
import sp.it.util.collections.materialize
import sp.it.util.reactive.Subscribed
import sp.it.util.reactive.Subscription
import sp.it.util.system.Os
import sp.it.util.units.millis
import java.io.File
import java.io.IOException
import java.nio.file.ClosedWatchServiceException
import java.nio.file.FileSystems
import java.nio.file.Path
import java.nio.file.StandardWatchEventKinds.ENTRY_CREATE
import java.nio.file.StandardWatchEventKinds.ENTRY_DELETE
import java.nio.file.StandardWatchEventKinds.ENTRY_MODIFY
import java.nio.file.StandardWatchEventKinds.OVERFLOW
import java.nio.file.WatchEvent
import java.nio.file.WatchEvent.Kind
import java.nio.file.WatchKey
import java.nio.file.WatchService
import java.util.concurrent.ConcurrentHashMap
import java.util.function.Predicate

/**
 *
 * Directory/File monitor. Supports filtering events, recursive monitoring and also attempts to reduce duplicate events.
 * The exact implementation may differ based on the type of desired monitoring and OS support. There are no guarantees
 * about the exact event timing or delays, event deduplication, performance or even event capture as [WatchService] is
 * known to fail to observe events in some scenarios.
 *
 * It is not simple to monitor files (in a cross platform way):
 *
 * - target
 * [WatchService] allows us to only monitor a directory.
 *
 * Solvable using a predicate parameter to filter out unwanted events.
 * - Modification events.
 *
 * Events can fire multiple times when application uses safe-rewrite saving or in various other scenarios. OSes also
 * behave differently and some fire MODIFY on CREATE too.
 * Solvable using event reducer & checking modification times
 * - Recursive monitoring
 *
 * CREATE and DELETE events are fired for files and directories up to level 2 (children of
 * children of the monitored directory, e.g.,  mon_dir/lvl1/file.txt). MODIFIED events will
 * be thrown for any direct child.
 *
 * Furthermore, recursive monitoring is only supported on Windows, this is a platform limitation.
 */
class FileMonitor {
    private lateinit var monitoredFileDir: File
    private lateinit var watchService: WatchService
    private lateinit var filter: Predicate<File>
    private lateinit var action: (Kind<Path>, File) -> Unit
    private val eventAccumulations = ConcurrentHashMap<File, Event>()
    private val eventAccumulating = Subscribed { startCleanup() }
    private val modificationReducer = EventReducer.thatConsumes<Event> { e ->
        when (e.kind) {
            ENTRY_MODIFY -> eventAccumulations[e.file] = e
            ENTRY_DELETE -> {
                eventAccumulations.remove(e.file)?.emit()
                e.emit()
            }
            else -> e.emit()
        }
    }

    private fun Event.emit() = runFX { action(kind, file) }

    private fun startCleanup(): Subscription {
        val timeEvict = 100
        val t = fxTimer(timeEvict.millis, -1) {
            val timeMs = System.currentTimeMillis()
            val files = eventAccumulations.keys.materialize()
            val events = files.mapNotNull { f ->
                eventAccumulations[f]
                        ?.takeIf { timeMs-it.timeMs>timeEvict }
                        ?.let { eventAccumulations.remove(f) }
            }
            events.forEach { it.emit() }
        }
        runFX { t.start() }

        return Subscription {
            eventAccumulations.values.forEach { it.emit() }
            t.stop()
        }
    }

    fun stop() {
        try {
            watchService.close()
        } catch (e: IOException) {
            logger.error(e) { "Error when closing file monitoring $monitoredFileDir" }
        }
        eventAccumulating.unsubscribe()
    }

    private class Event constructor(kind: Kind<Path>, file: File) {
        val file = file
        val kind = kind
        val timeMs = System.currentTimeMillis()
    }

    companion object: KLogging() {

        /**
         * Creates and starts directory monitoring reporting events for single file contained within the directory.
         *
         * Only events for the specified file will be provided.
         *
         * @param monitoredFile file to be monitored
         * @param action handles the event on fx application thread
         * @return directory monitor
         */
        fun monitorFile(monitoredFile: File, action: (Kind<Path>) -> Unit) = monitorDirectory(monitoredFile.parentFile!!, false, { monitoredFile==it }, { type, _ -> action(type) })

        /**
         * Creates and starts directory monitoring for specified directory reporting events for any 1st
         * level child file which passes the predicate filter.
         *
         * Events for the specified directory and its children (up to depth 1 or unlimited) will be provided.
         *
         * @param monitoredDir directory to be monitored
         * @param recursive determines whether events for files in depth more than 1 will be provided
         * @param filter filter narrowing down events, for example any text file
         * @param action handles the event on fx application thread
         * @return directory monitor
         */
        fun monitorDirectory(monitoredDir: File, recursive: Boolean, filter: (File) -> Boolean = { true }, action: (Kind<Path>, File) -> Unit): FileMonitor {
            val fm = FileMonitor()
            fm.monitoredFileDir = monitoredDir
            fm.action = action
            fm.eventAccumulating.subscribe()

            try {
                fm.watchService = FileSystems.getDefault().newWatchService()

                runNew {
                    fun reg(p: File) {
                        try {
                            when {
                                Os.WINDOWS.isCurrent -> {
                                    if (recursive) p.toPath().register(fm.watchService, arrayOf(ENTRY_CREATE, ENTRY_DELETE, ENTRY_MODIFY, OVERFLOW), FILE_TREE)
                                    else p.toPath().register(fm.watchService, ENTRY_CREATE, ENTRY_DELETE, ENTRY_MODIFY, OVERFLOW)
                                }
                                else -> {
                                    fun reg1(f: File) = f.toPath().register(fm.watchService, ENTRY_CREATE, ENTRY_DELETE, ENTRY_MODIFY, OVERFLOW)
                                    fun regN(f: File) = f.walkTopDown().materialize().onEach { if (it.isDirectory) reg1(it) }

                                    if (recursive) regN(p)
                                    else reg1(p)
                                }
                            }
                        } catch (e: IOException) {
                            logger.error(e) { "Failed to register file watch for $p" }
                        } catch (e: ClosedWatchServiceException) {
                            // watcher is already closed, logging this would pollute log with error-like non errors
                        }
                    }

                    reg(fm.monitoredFileDir)

                    threadFactory("FileMonitor-${monitoredDir.path}", true).newThread {
                        var valid: Boolean
                        var watchKey: WatchKey
                        do {
                            try {
                                watchKey = fm.watchService.take()
                            } catch (e: InterruptedException) {
                                logger.error(e) { "Interrupted monitoring of directory ${fm.monitoredFileDir}" }
                                return@newThread
                            } catch (e: ClosedWatchServiceException) {
                                return@newThread
                            }

                            @Suppress("UNCHECKED_CAST")
                            for (event in watchKey.pollEvents()) {
                                val type = event.kind()
                                if (type!==OVERFLOW) {
                                    val ev = event as WatchEvent<Path>
                                    val modifiedFileName = ev.context().toString()
                                    val modifiedFile = File(fm.monitoredFileDir, modifiedFileName)

                                    if (type===ENTRY_CREATE && !Os.WINDOWS.isCurrent && modifiedFile.isDirectory)
                                        reg(modifiedFile)

                                    if (filter(modifiedFile))
                                        fm.modificationReducer.push(Event(type as Kind<Path>, modifiedFile))
                                }
                            }

                            valid = watchKey.reset()
                        } while (valid)
                    }.start()
                }

            } catch (e: IOException) {
                logger.error(e) { "Error when starting directory monitoring ${fm.monitoredFileDir}" }
            }


            return fm
        }

    }

}